sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DefaultColumns.scala

```diff
@@ -0,0 +1,287 @@
+                  case u: UnresolvedAttribute if u.name.equalsIgnoreCase(default) => defaultExpr
+                  case _ => expr
+                }
+            }
+          }
+        table.copy(rows = newRows)
+      case project: Project if project.projectList.size == defaultExprs.size =>
+        val updated: Seq[NamedExpression] =
+          // Map each expression of the project list to its corresponding DEFAULT expression in the
+          // INSERT INTO object table, if the two lists are equal in length.
+          (project.projectList, defaultExprs, colNames).zipped.map {
+            case (expr: Expression, defaultExpr: Expression, colName: String) =>
+              expr match {
+                case u: UnresolvedAttribute if u.name.equalsIgnoreCase(default) =>
+                  Alias(defaultExpr, colName)()
+                case _ => expr
+              }
+          }
+        project.copy(projectList = updated)
+      case _ => insert.query
+    }
+    insert.copy(query = newQuery)
+  }
+
+  /**
+   * Parses DEFAULT column text to an expression, returning a reasonable error upon failure.
+   *
+   * @param colName the name of the DEFAULT column whose text we endeavor to parse.
+   * @param colText the string contents of the DEFAULT column value.
+   * @param statementType which type of statement we are running, such as INSERT; useful for errors.
+   * @return the expression resulting from the parsing step.
+   */
+  private def Parse(colName: String, colText: String, statementType: String): Expression = {
+    try {
+      parser.parseExpression(colText)
+    } catch {
+      case ex: ParseException =>
+        throw new AnalysisException(
+          s"Failed to execute $statementType command because the destination table column " +
+            colName + analysisPrefix + s"$colText yields ${ex.getMessage}")
+    }
+  }
+
+  /**
+   * Analyzes and constant-folds `colExpr`, returning a reasonable error message upon failure.
+   *
+   * @param colExpr result of a parsing operation suitable for consumption by analysis.
+   * @param colText string contents of the DEFAULT column value; useful for errors.
+   * @param colName string name of the DEFAULT column; useful for errors.
+   * @param statementType which type of statement we are running, such as INSERT; useful for errors.
+   * @return Result of the analysis and constant-folding operation.
+   */
+  private def Analyze(colExpr: Expression, colText: String, colName: String,
+      statementType: String):
+  Expression = {
+    try {
+      // Invoke the analyzer over the 'colExpr'.
+      val plan = analyzer.execute(Project(Seq(Alias(colExpr, colName)()), OneRowRelation()))
+      analyzer.checkAnalysis(plan)
+      // Perform constant folding over the result.
+      val folded = ConstantFolding(plan)
+      val result = folded match {
+        case Project(Seq(a: Alias), OneRowRelation()) => a.child
+      }
+      // Make sure the constant folding was successful.
```

Comment:
Shall we also check the result data type is compatible with the column type? E.g. `i int default 'abc'`